### PR TITLE
Add support for kotlin suspend function

### PR DESCRIPTION
**Pull request by apoupard**
Contribution for #634 

Add support for:
Function: suspend (Flow<Person>) -> Flow<String>
Consumer: suspend (Flow<String>) -> Unit 
Supplier(): suspend () -> Flow<String>

### CODE REVIEW

**Comment by pivotal-issuemaster** [comment]
@apoupard Please sign the [Contributor License Agreement](https://cla.pivotal.io/sign/spring?repositoryId=spring-cloud/spring-cloud-function&pullRequestId=655)!

[Click here](https://cla.pivotal.io/sync/spring?repositoryId=spring-cloud/spring-cloud-function&pullRequestId=655) to manually synchronize the status of this Pull Request.

See the [FAQ](https://cla.pivotal.io/about) for frequently asked questions.

**Comment by pivotal-issuemaster** [comment]
@apoupard Thank you for signing the [Contributor License Agreement](https://cla.pivotal.io/sign/spring?repositoryId=spring-cloud/spring-cloud-function&pullRequestId=655)!

**Comment by olegz** [comment]
@apoupard Thank you so much for your contribution. I will review shortly and let you know if additional work is needed or merge it.

**Comment by olegz** [comment]
Also, do you think this PR addresses #634 fully or there are more work to come?

**Comment by apoupard** [comment]
Technical implementation of `suspend (Flow) -> Flow`, Consumer:` suspend (Flow) -> Unit` and Supplier(): `suspend () -> Flow` is fully functional with web.
I didn't test them with rsocket and Faas providers.
But I didn't do any change on the documentation. I'm not event sure where documentation sources are.

For non coroutine lambda, I noticed Consumer: (String) -> Unit is mapped as Function<String, Unit> instead of as Consumer<String>, but it's should prob be done in another PR.

Support could be add for:
```
suspend (String) -> String
suspend (String) -> Unit
suspend (Unit) -> String
```
 But as lamdba like `Function<Mono<String>, Mono<String>>, ...` are not supported in java, simple suspend lambda doesn't look to be  in the state of mind of spring-cloud-function.

**Comment by olegz** [comment]
That's fine, if there are any issues with integrations such as web, rsocket etc, we can address them individually

**Comment by olegz** [comment]
And doc sources are [here](https://github.com/spring-cloud/spring-cloud-function/blob/master/docs/src/main/asciidoc/spring-cloud-function.adoc)

File: spring-cloud-function-kotlin/src/main/java/org/springframework/cloud/function/context/config/KotlinLambdaToFunctionAutoConfiguration.java [hunk 1]
```diff
@@ -49,6 +37,14 @@
...
 import org.springframework.core.ResolvableType;
 import org.springframework.util.ObjectUtils;
+import reactor.core.publisher.Flux;
+
+import java.lang.reflect.ParameterizedType;
+import java.lang.reflect.Type;
+import java.util.function.Consumer;
+import java.util.function.Function;
+import java.util.function.Supplier;
+
```
Comment (olegz) [inline spring-cloud-function-kotlin/src/main/java/org/springframework/cloud/function/context/config/KotlinLambdaToFunctionAutoConfiguration.java:47]: Please add your name as `@author` (and other classes you touched)

File: spring-cloud-function-kotlin/src/main/java/org/springframework/cloud/function/context/config/KotlinLambdaToFunctionAutoConfiguration.java [hunk 2]
```diff
@@ -16,21 +16,9 @@
...
-import java.util.function.Function;
-import java.util.function.Supplier;
-
-
-import kotlin.jvm.functions.Function0;
-import kotlin.jvm.functions.Function1;
-import kotlin.jvm.functions.Function2;
-import kotlin.jvm.functions.Function3;
-import kotlin.jvm.functions.Function4;
+import kotlin.jvm.functions.*;
```
Comment (olegz) [inline spring-cloud-function-kotlin/src/main/java/org/springframework/cloud/function/context/config/KotlinLambdaToFunctionAutoConfiguration.java:19]: We generally do not allow * imports, so the build actually fails with checkstyle violations. 
Just run `./mvnw clean install` and you'll see the violations. Our formatting details and plugins for intellij and eclipse are [here](https://github.com/spring-cloud/spring-cloud-build#code-conventions-and-housekeeping)

**Comment by olegz** [comment]
After fixing the checkstyles, so I can build I am getting compilation errors when doing a full build

```
[ERROR] Failed to execute goal org.apache.maven.plugins:maven-compiler-plugin:3.8.1:compile (java-compile) on project spring-cloud-function-kotlin: Compilation failure: Compilation failure: 
[ERROR] /Users/ozhurakousky/dev/repo/spring-cloud-function/spring-cloud-function-kotlin/src/main/java/org/springframework/cloud/function/context/config/KotlinLambdaToFunctionAutoConfiguration.java:[159,44] cannot find symbol
[ERROR]   symbol:   method isValidSuspendingFunction(java.lang.Object,java.lang.Object)
[ERROR]   location: class org.springframework.core.CoroutinesUtils
[ERROR] /Users/ozhurakousky/dev/repo/spring-cloud-function/spring-cloud-function-kotlin/src/main/java/org/springframework/cloud/function/context/config/KotlinLambdaToFunctionAutoConfiguration.java:[160,81] incompatible types: java.lang.Object cannot be converted to java.lang.reflect.Method
[ERROR] /Users/ozhurakousky/dev/repo/spring-cloud-function/spring-cloud-function-kotlin/src/main/java/org/springframework/cloud/function/context/config/KotlinLambdaToFunctionAutoConfiguration.java:[167,44] cannot find symbol
[ERROR]   symbol:   method isValidSuspendingSupplier(java.lang.Object)
[ERROR]   location: class org.springframework.core.CoroutinesUtils
[ERROR] /Users/ozhurakousky/dev/repo/spring-cloud-function/spring-cloud-function-kotlin/src/main/java/org/springframework/cloud/function/context/config/KotlinLambdaToFunctionAutoConfiguration.java:[168,55] cannot find symbol
. . .
```

File: spring-cloud-function-kotlin/src/main/java/org/springframework/cloud/function/context/config/KotlinLambdaToFunctionAutoConfiguration.java (hunk 1 above)
Comment (apoupardsmartb) [reply spring-cloud-function-kotlin/src/main/java/org/springframework/cloud/function/context/config/KotlinLambdaToFunctionAutoConfiguration.java:52]: Done

File: spring-cloud-function-kotlin/src/main/java/org/springframework/cloud/function/context/config/KotlinLambdaToFunctionAutoConfiguration.java (hunk 2 above)
Comment (apoupardsmartb) [reply spring-cloud-function-kotlin/src/main/java/org/springframework/cloud/function/context/config/KotlinLambdaToFunctionAutoConfiguration.java]: Done

**Comment by apoupardsmartb** [comment]
@olegz I fixed compilation error and `./mvnw clean install` is running successfully.

**Comment by apoupard** [comment]
And about the error during the travis build, I had a similar issue on my host, I had to clone and build the project spring-cloud-release in order to have spring-cloud-dependencies in my local repo.

**Comment by olegz** [comment]
I’ll review it tomorrow

**Comment by olegz** [comment]
@apoupard Once again, thank you for your contribution.
It would be nice if you could also contribute to the [Kotlin section](https://docs.spring.io/spring-cloud-function/docs/3.1.1/reference/html/spring-cloud-function.html#_kotlin_lambda_support) of the documentation. [Here is](https://github.com/spring-cloud/spring-cloud-function/blob/master/docs/src/main/asciidoc/spring-cloud-function.adoc#kotlin-lambda-support) the actual adoc file.

Cheers

**Comment by apoupard** [comment]
Ok I will work on the documentation tonighth.

**Pull request closed without merging**